sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -86,6 +88,56 @@ case class Mode(
     buffer
   }
 
+  private def getCollationAwareBuffer(
+      childDataType: DataType,
+      buffer: OpenHashMap[AnyRef, Long]): Iterable[(AnyRef, Long)] = {
+    def groupAndReduceBuffer(groupingFunction: AnyRef => _): Iterable[(AnyRef, Long)] = {
+      buffer.groupMapReduce(t =>
+        groupingFunction(t._1))(x => x)((x, y) => (x._1, x._2 + y._2)).values
+    }
+    def determineBufferingFunction(
+        childDataType: DataType): Option[AnyRef => _] = {
+      childDataType match {
+        case _ if UnsafeRowUtils.isBinaryStable(child.dataType) => None
+        case _ => Some(collationAwareTransform(_, childDataType))
+      }
+    }
+    determineBufferingFunction(childDataType).map(groupAndReduceBuffer).getOrElse(buffer)
+  }
+
+  private def collationAwareTransform(myData: AnyRef, myElementType: DataType): AnyRef = {
+    myElementType match {
+      case _ if UnsafeRowUtils.isBinaryStable(myElementType) => myData
+      case myStructType: StructType =>
+        processStructTypeWithBuffer(
+          myData.asInstanceOf[InternalRow].toSeq(
+            myStructType).zip(
+            myStructType.fields))
+      case myArrayType: ArrayType => processArrayTypeWithBuffer(
+        myArrayType,
+        myData.asInstanceOf[ArrayData])
+      case myStringType: StringType =>
+        CollationFactory.getCollationKey(
+          myData.asInstanceOf[UTF8String],
+          myStringType.collationId)
+      case _ =>
+        throw new UnsupportedOperationException(
+          s"Unsupported data type for collation-aware mode: $myElementType")
+    }
+  }
```

Comment:
```suggestion
  private def collationAwareTransform(data: AnyRef, dataType: DataType): AnyRef = {
    dataType match {
      case _ if UnsafeRowUtils.isBinaryStable(dataType) =>
        data
      case st: StructType =>
        processStructTypeWithBuffer(data.asInstanceOf[InternalRow].toSeq(st).zip(st.fields))
      case at: ArrayType =>
        processArrayTypeWithBuffer(at, data.asInstanceOf[ArrayData])
      case st: StringType =>
        CollationFactory.getCollationKey(data.asInstanceOf[UTF8String], st.collationId)
      case _ =>
        throw new UnsupportedOperationException(
          s"Unsupported data type for collation-aware mode: $dataType")
    }
  }
```
let's try to keep the code structure and naming patterns consistent with the codebase